include/lib/reentry_guard.hpp

```diff
@@ -0,0 +1,91 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <thread>
+
+namespace ddprof {
+
+class ThreadEntries {
+public:
+  static constexpr size_t max_threads = 10;
+  std::array<std::atomic<pid_t>, max_threads> thread_entries;
+  ThreadEntries() { reset(); }
+  void reset() {
+    for (auto &entry : thread_entries) {
+      entry.store(-1, std::memory_order_relaxed);
+    }
+  }
+};
+
+class TLReentryGuard {
+public:
+  explicit TLReentryGuard(ThreadEntries &entries, pid_t tid)
+      : _entries(entries), _tid(tid), _ok(false), _index(-1) {
+    while (true) {
+      for (size_t i = 0; i < ThreadEntries::max_threads;
+           ++i) {
+        pid_t expected = -1;
+        if (_entries.thread_entries[i].compare_exchange_strong(
```

Comment:
Minor: I think a compare_exchange_weak could also work considering we would try the next entry or loop.